js/packages/openinference-instrumentation-llama-index/src/instrumentation.ts

```diff
@@ -87,72 +98,160 @@ export class LlamaIndexInstrumentation extends InstrumentationBase<
+    ) {
+      const span = instrumentation.tracer.startSpan(`query`, {
+        kind: SpanKind.INTERNAL,
+        attributes: {
+          [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+            OpenInferenceSpanKind.CHAIN,
+          [SemanticConventions.INPUT_VALUE]: args[0].query,
+        },
+      });
+
+      const execContext = getExecContext(span);
+
+      const execPromise = safeExecuteInTheMiddle<
+        ReturnType<typeof module.RetrieverQueryEngine.prototype.query>
+      >(
+        () => {
+          return context.with(execContext, () => {
+            return original.apply(this, args);
+          });
+        },
+        (error) => {
+          // Push the error to the span
+          if (error) {
+            span.recordException(error);
+            span.setStatus({
+              code: SpanStatusCode.ERROR,
+              message: error.message,
+            });
+            span.end();
+          }
+        },
+      );
+
+      const wrappedPromise = execPromise.then((result) => {
+        span.setAttributes({
+          [SemanticConventions.OUTPUT_VALUE]: result.response,
+        });
+        span.end();
+        return result;
+      });
+      return context.bind(execContext, wrappedPromise);
+    };
+  }
+
+  private patchRetrieveMethod(
+    original: typeof module.VectorIndexRetriever.prototype.retrieve,
+    module: typeof llamaindex,
+    instrumentation: LlamaIndexInstrumentation
+  ) {
+    return function patchedRetrieve(
+      this: unknown,
+      ...args: Parameters<typeof module.VectorIndexRetriever.prototype.retrieve>
+    ) {
+
+      // Start the span with initial attributes: kind, span_kind, input
+      const span = instrumentation.tracer.startSpan(`retrieve`, {
+        kind: SpanKind.INTERNAL,
+        attributes: {
+          [SemanticConventions.OPENINFERENCE_SPAN_KIND]:
+            OpenInferenceSpanKind.RETRIEVER,
+          [SemanticConventions.INPUT_VALUE]: args[0].query,
+        },
+      });
+
+      const execContext = getExecContext(span);
+
+      const execPromise = safeExecuteInTheMiddle<
+        ReturnType<typeof module.VectorIndexRetriever.prototype.retrieve>
+      >(
+        () => {
+          return context.with(execContext, () => {
+            return original.apply(this, args);
+          });
+        },
+        (error) => {
+          // Push the error to the span
+          if (error) {
+            span.recordException(error);
+            span.setStatus({
+              code: SpanStatusCode.ERROR,
+              message: error.message,
+            });
+            span.end();
+          }
+        },
+      );
+
+      const wrappedPromise = execPromise.then((result) => {
+        const docs: Attributes = {};
+        result.forEach((document, index) => {
+          const { node, score } = document;
+
+          if (node instanceof TextNode) {
+            const textNode = node as TextNode;
```

Comment:
this cast shouldn't be necesary here given the check on line 237 above, curious if it could be removed?